read APBS .dx.gz generated grids

eloyfelix opened the pull request:
APBS can write gz compressed dx files: https://apbs-pdb2pqr.readthedocs.io/en/latest/apbs/input/elec/write.html
It's quite convenient when dealing with very big grids as it needs less disk space/RAM.

This PR adds DXGZ input format functionality.

eloyfelix commented:
Travis shows 'pending' state in github but it already finished.

This could be used like this:
```python
from gridData import Grid

grid = Grid("data.dx.gz", file_format='DXGZ')
```

Is this something that you would merge into master?

Thanks

[data.dx.gz](https://github.com/MDAnalysis/GridDataFormats/files/3797657/data.dx.gz)

kain88-de commented:
Hi @eloyfelix this looks like a nice addition. Definitely good for merging. Can you add a test for this and add yourself to the authors?

I wonder if we should just allow this for all files. If it ends in `.gz` we just assume it is zipped.

orbeckst commented:
The functionality is very useful – not sure why I didn't put it in the first place because I gzip my dx files, too.

But I don't like additional kwargs to indicate compression. I agree with @kain88-de that this should be universal and should just be based on the extension.
> I wonder if we should just allow this for all files. If it ends in .gz we just assume it is zipped.

We have functionality in MDAnalysis `openany()` https://github.com/MDAnalysis/mdanalysis/blob/7f82b88f769f3882e44b5121463c93b899ef5b55/package/MDAnalysis/lib/util.py#L265-L454 to do this seamlessly. The problem is that MDA is GPL2 and GridDataFormats is LGPL. My understanding is that we cannot copy code from GPL to LGPL. Instead I have to find out who wrote the code, get everyone who contributed to agree to relicense, and then we can take it.

orbeckst commented:
@eloyfelix you can either wait and see if we can just take the `openany()` code, which will just solve the problem, or if you are keen to move forward, change your code so that it looks at the filename extension to decide if it should uncompress or compress. In this way, the API will stay the same even if we add `openany()` later.

orbeckst commented:
# Relicensing of `openany()` under LGPL

From [blame: lib/util.py L265-L454](https://github.com/MDAnalysis/mdanalysis/blame/develop/package/MDAnalysis/lib/util.py#L265-L454): Do you agree to relicense `openany()` (see link for exact lines of code) under LGPL? Please reply below (you can also tick the box but there must be a reply with your handle in this thread):
- [x] @orbeckst 
- [ ] @jbarnoud 
- [ ] @richardjgowers 
- [ ] @utkbansal 
- [ ] @kain88-de 
- [ ] @tylerjereddy 

Thanks!

orbeckst commented on gridData/OpenDX.py:
```diff
@@ -483,7 +484,7 @@ def write(self, filename):
                 outfile.write('component "%s" value %s\n' % (component,str(object.id)))
 
-    def read(self,file):
+    def read(self, file, gz=False):
```
Do not use a kwarg but detect compression from the filename.

orbeckst commented on gridData/OpenDX.py:
```diff
@@ -492,7 +493,7 @@ def read(self,file):
         DXfield = self
         p = DXParser(file)
-        p.parse(DXfield)
+        p.parse(DXfield, gz)
```
Avoid the gz flag.

orbeckst commented on gridData/OpenDX.py:
```diff
@@ -678,8 +679,13 @@ def parse(self,DXfield):
         self.tokens = []                    # token buffer
-        with open(self.filename, 'r') as self.dxfile:
-            self.use_parser('general')      # parse the whole file and populate self.objects
+
```
Do something like
```python
if filename.endswith('.gz'):
   with gzip.open(self.filename, 'rt') as self.dxfile:
                self.use_parser('general')
else:
    with open(self.filename, 'r') as self.dxfile:
                self.use_parser('general')      # parse the whole file and populate self.objects
```

Once we add `openany()` to the code, this can be replaced with
```python
with openany(filename):
     self.use_parser('general')
```

orbeckst commented on gridData/core.py:
```diff
@@ -118,13 +119,15 @@ def __init__(self, grid=None, edges=None, origin=None, delta=None,
         # file formats are guess from extension == lower case key
         self._exporters = {
             'DX': self._export_dx,
+            'DXGZ': self._export_dxgz,
```
No extra format, detect by file extension.

orbeckst commented on gridData/core.py:
```diff
@@ -118,13 +119,15 @@ def __init__(self, grid=None, edges=None, origin=None, delta=None,
         self._loaders = {
             'CCP4': self._load_cpp4,
             'DX': self._load_dx,
+            'DXGZ': self._load_dxgz,
```
No extra format, detect by file extension.

orbeckst commented on gridData/core.py:
```diff
@@ -432,13 +435,17 @@ def _load_cpp4(self, filename):
         self.__init__(grid=grid, edges=edges, metadata=self.metadata)
 
-    def _load_dx(self, filename):
+    def _load_dx(self, filename, gz=False):
```
No `gz=...`, detect by file extension.

orbeckst commented on gridData/core.py:
```diff
@@ -546,6 +553,18 @@ def _export_dx(self, filename, type=None, typequote='"', **kwargs):
         dx = OpenDX.field('density', components=components, comments=comments)
         dx.write(filename)
 
+    def _export_dxgz(self, filename, type=None, typequote='"', **kwargs):
```
Too much code duplication, use the filename extension logic in `_export_dx()`. (Can later be replaced with `openany(filename, "w")`.)

orbeckst commented on gridData/tests/datafiles/__init__.py:
```diff
@@ -2,9 +2,10 @@
 from pkg_resources import resource_filename
 
-__all__ = ["DX", "CCP4", "gOpenMol"]
+__all__ = ["DX", "DXGZ", "CCP4", "gOpenMol"]
```
no extra DXGZ format.

orbeckst commented on gridData/tests/test_dx.py:
```diff
@@ -21,6 +21,16 @@ def test_read_dx():
     assert_equal(g.delta, np.ones(3))
     assert_equal(g.origin, np.array([20.1, 3., -10.]))
 
+def test_read_dxgz():
```
We should be able to just parametrize the previous test over the uncompressed and compressed file – reduced code duplication.

orbeckst commented on gridData/tests/test_dx.py:
```diff
@@ -66,6 +76,51 @@ def test_write_dx(tmpdir, nptype, dxtype, counts=100, ndim=3):
 
     assert_equal(out_dxtype, dxtype)
 
+@pytest.mark.parametrize("nptype,dxtype", [
```
Avoid code duplication, just parametrize the other test for the uncompressed file over the compressed and uncompressed files.

orbeckst commented on gridData/tests/test_dx.py:
```diff
@@ -80,5 +135,16 @@ def test_write_dx_ValueError(tmpdir, nptype, counts=100, ndim=3):
             outfile = "grid.dx"
             g.export(outfile)
 
+@pytest.mark.parametrize('nptype', ("complex64", "complex128", "bool_"))
```
As above: avoid code duplication, use parametrized tests.

eloyfelix commented:
Many thanks for your suggestions @orbeckst. I'll take a look at it asap and try to make the changes to better fit your existing codebase.

eloyfelix commented:
this should be, at least, a bit closer :P

orbeckst commented on gridData/tests/test_dx.py:
```diff
@@ -10,19 +10,9 @@
-
-def test_read_dxgz():
-    g = Grid(datafiles.DXGZ, file_format='DXGZ')
+@pytest.mark.parametrize("infile", [datafiles.DX, datafiles.DX+'.gz'])
```
I think you also need to
- add a `test.dx.gz` to test files
- add `DXGZ` to `datafiles/__init__.py`

I don't know if one can make a fixture work in which you generate the gzipped file on the fly. I can think of the first part, but not how to use the fixture together with `datafiles.DX` in a parameterized test or fixture. But in any case, here's the code I can think of (uses [tmpdir_factory](https://docs.pytest.org/en/latest/tmpdir.html#the-tmpdir-factory-fixture)  because the temporary file should be created in a temporary location but the `tmpfile` fixture is not module level)
```python
@pytest.fixture(scope="module")
def dxgz_file(tmpdir_factory, src=datafiles.DX):
    basename = os.path.basename(src) + '.gz'
    fn = tmpdir_factory.mktemp("compressed").join(basename)
    with open(src, 'rb') as inp:
        with gzip.open(str(fn), "wb") as out:
             out.write(inp.read())
    return str(fn)

# use dxgz_file fixture when needed ... although I am actually not sure how 
# you use a fixture as part of a parameterized fixture.
```

eloyfelix replied:
I added test.dx.gz file and DXGZ to datafiles/__init__.py but I'm not sure about the fixture...

orbeckst commented on gridData/tests/test_dx.py:
```diff
@@ -21,7 +22,7 @@ def test_read_dx():
     assert_equal(g.origin, np.array([20.1, 3., -10.]))
 
-
+@pytest.mark.parametrize("outfile", ["grid.dx", "grid.dx.gz"])
```
good!

eloyfelix replied:
cool!

utkbansal commented:
I'm okay with the changes.

eloyfelix commented on CHANGELOG:
```diff
@@ -13,6 +13,15 @@ The rules for this file:
+
+  Enhancements
+
+  * Allow parsing/writting gzipped DX files
```
done! But you definitely did stuff for this release :P

orbeckst replied:
Nope, I didn't. I just checked. (Reviewing PRs does not count for me.)

eloyfelix commented on gridData/core.py:
```diff
@@ -552,18 +550,12 @@ def _export_dx(self, filename, type=None, typequote='"', **kwargs):
-                for l in in_file:
-                    out_file.write(l)
-        os.remove(filename)
+        if ext == '.gz':
```
I moved the logic to dx.write but given how it was originally implemented it was a bit tricky. I needed to add few isinstance to check if I needed to write string or bytes depending on the stream

eloyfelix replied:
Also needed to implement a simple version of your 'openany' to deal with it:

https://github.com/eloyfelix/GridDataFormats/blob/DXGZ_input_format/gridData/OpenDX.py#L486

orbeckst replied:
I see. I don't particularly like having to repeat the same code everywhere. But that's because of the poor structure of the original code (i.e., my fault ;-)  where writing is actually done in each element of the DXfield. Can you think of a better code structure where we only need to do the encoding etc once? E.g., instead of having each element write, return a text representation that is then concatenated at the top level.

eloyfelix replied:
added write_line function: https://github.com/eloyfelix/GridDataFormats/blob/DXGZ_input_format/gridData/OpenDX.py#L193

It's probably not optimal but it's the best that I could do without doing major codebase changes (which I'm not sure I'm ready to do)

orbeckst replied:
Your approach is fine, I just added a few minor change requests to clean up the code. Importantly, this is really cleaning up my mess – I would be very grateful if you could do it as part of this PR!

eloyfelix commented:
hope we are getting closer with the latest changes

orbeckst commented on gridData/OpenDX.py:
```diff
@@ -166,6 +166,7 @@
 from six import next
 from six.moves import range
 import gzip
+import bz2
```
If you allow bz2 then that also needs to be available when writing.

kain88-de replied:
Since we do not have code to read bz2 files right now, at least I can only find code to read .gz files, I would recommend we leave that out of this PR. @eloyfelix If you like you can later open a second PR to add the bz2.

eloyfelix replied:
makes sense, I think that I'll remove it for now...

orbeckst commented on gridData/OpenDX.py:
```diff
@@ -460,6 +483,13 @@ def __init__(self,classid='0',components=None,comments=None):
 
+    def _openfile(self, filename):
+        """Returns a regular or gz file descriptor"""
+        if filename.endswith('.gz'):
```
If you allow bz2 then that also needs to be available when writing.

eloyfelix replied:
makes sense, I think that I'll remove it for now...

kain88-de commented:
LGTM. @orbeckst I leave you to merge it

@eloyfelix Nice work. Thanks

orbeckst commented on gridData/OpenDX.py:
```diff
@@ -385,12 +397,13 @@ def write(self, file):
                 for i in range(values_per_line):
-                    file.write(fmt_string.format(next(values)) + "\t")
-                file.write('\n')
+                    DXclass.write_line(
```
same here: `self._write_line(file, ...)`

eloyfelix replied:
done

orbeckst commented on gridData/OpenDX.py:
```diff
@@ -385,12 +397,13 @@ def write(self, file):
-                file.write('\n')
+                    DXclass.write_line(
+                        self, file, fmt_string.format(next(values)) + "\t")
+                DXclass.write_line(self, file, '\n')
```
same here: `self._write_line(file, ...)`

eloyfelix replied:
done

orbeckst commented on gridData/OpenDX.py:
```diff
@@ -227,12 +236,15 @@ def __init__(self,classid,shape=None,origin=None,delta=None,**kwargs):
+    def write(self, file):
+        DXclass.write(self, file, ('counts '+self.ndformat(' %d')) %
+                      tuple(self.shape))
+        DXclass.write_line(self, file, 'origin %f %f %f\n' %
```
same here: `self._write_line(file, ...)`

eloyfelix replied:
done

orbeckst commented on gridData/OpenDX.py:
```diff
@@ -227,12 +236,15 @@ def __init__(self,classid,shape=None,origin=None,delta=None,**kwargs):
+                           tuple(self.origin))
         for delta in self.delta:
-            file.write(('delta '+self.ndformat(' %f')+'\n') % tuple(delta))
+            DXclass.write_line(
```
same here: `self._write_line(file, ...)`

eloyfelix replied:
done

orbeckst commented on gridData/OpenDX.py:
```diff
@@ -459,6 +472,13 @@ def __init__(self,classid='0',components=None,comments=None):
         self.components = components
         self.comments= comments
 
+    def _openfile(self, filename):
```
Rename to `_openfile_writing()` or similar to make sure that it is clear that this _only_ opens a file for writing. Otherwise one immediately wonders where the `"w"` kwarg is...

eloyfelix replied:
done!

orbeckst commented on gridData/OpenDX.py:
```diff
@@ -459,6 +472,13 @@ def __init__(self,classid='0',components=None,comments=None):
         self.comments= comments
 
+    def _openfile(self, filename):
+        """Returns a regular or gz file descriptor"""
```
Say in docs that the file is opened for writing.

eloyfelix replied:
done!

orbeckst commented on gridData/OpenDX.py:
```diff
@@ -471,19 +491,20 @@ def write(self, filename):
         # comments (VMD chokes on lines of len > 80, so truncate)
         maxcol = 80
-        with open(str(filename), 'w') as outfile:
+        with self._openfile(str(filename)) as outfile:
```
change name, see above

eloyfelix replied:
done

orbeckst commented on gridData/OpenDX.py:
```diff
@@ -471,19 +491,20 @@ def write(self, filename):
-            DXclass.write(self,outfile,quote=True)
-            for component,object in self.sorted_components():
-                outfile.write('component "%s" value %s\n' % (component,str(object.id)))
+            DXclass.write(self, outfile, quote=True)
```
use `super`

orbeckst commented on gridData/core.py:
```diff
@@ -32,6 +32,7 @@
 
 import os
 import errno
+import gzip
```
remove unused import

eloyfelix replied:
done

orbeckst commented on gridData/tests/test_dx.py:
```diff
@@ -10,8 +10,9 @@
 
-def test_read_dx():
-    g = Grid(datafiles.DX)
+@pytest.mark.parametrize("infile", [datafiles.DX, datafiles.DXGZ])
```
This is good, totally happy with having the gzipped file as part of the tests. It's so tiny, it doesn't matter.

eloyfelix replied:
cool!

eloyfelix commented:
Thanks for the review, I believe I did all the proposed changes. The whole thing looks definitely cleaner now.
Hope to see a release soon! :P

eloyfelix commented on gridData/OpenDX.py:
```diff
@@ -471,19 +491,20 @@ def write(self, filename):
             for line in self.comments:
                 comment = '# '+str(line)
-                outfile.write(comment[:maxcol]+'\n')
+                DXclass.write_line(self, outfile, comment[:maxcol]+'\n')
```
done

eloyfelix commented on gridData/OpenDX.py:
```diff
@@ -471,19 +491,20 @@ def write(self, filename):
-                outfile.write('component "%s" value %s\n' % (component,str(object.id)))
+            DXclass.write(self, outfile, quote=True)
+            for component, object in self.sorted_components():
+                DXclass.write_line(self, outfile, 'component "%s" value %s\n' % (
```
done

eloyfelix commented on gridData/OpenDX.py:
```diff
@@ -177,15 +178,23 @@ def __init__(self,classid):
-        file.write('object '+classid+' class '+str(self.name)+' '+\
-                   optstring+'\n')
+        to_write = 'object '+classid+' class '+str(self.name)+' '+optstring+'\n'
+        if isinstance(file, gzip.GzipFile):
```
done!

eloyfelix commented:
Thanks for the review, I believe I did all the proposed changes. The whole thing looks definitely cleaner now.
Hope to see a release soon! :P

eloyfelix commented on gridData/OpenDX.py:
```diff
@@ -385,12 +397,13 @@ def write(self, file):
+                DXclass.write_line(self, file, '\n')
                 break
-        file.write('attribute "dep" string "positions"\n')
+        DXclass.write_line(self, file, 'attribute "dep" string "positions"\n')
```
done

eloyfelix commented on gridData/OpenDX.py:
```diff
@@ -385,12 +397,13 @@ def write(self, file):
+                DXclass.write_line(self, file, '\n')
             except StopIteration:
-                file.write('\n')
+                DXclass.write_line(self, file, '\n')
```
done

eloyfelix commented on gridData/OpenDX.py:
```diff
@@ -227,12 +236,15 @@ def __init__(self,classid,shape=None,origin=None,delta=None,**kwargs):
-                      ('counts '+self.ndformat(' %d')) % tuple(self.shape))
-        file.write('origin %f %f %f\n' % tuple(self.origin))
+    def write(self, file):
+        DXclass.write(self, file, ('counts '+self.ndformat(' %d')) %
```
done!

eloyfelix commented on gridData/OpenDX.py:
```diff
@@ -177,15 +178,23 @@ def __init__(self,classid):
+            to_write = to_write.encode()
+        file.write(to_write)
+
+    def write_line(self, file, line="", quote=False):
```
done

orbeckst commented:
Thank you for all the work and extra work addressing the review comments. This is a nice addition!

eloyfelix commented:
Always happy to contribute to nice libraries :)
Thanks for reviewing and merging.

El dj., 7 de nov. 2019, 16:53, Oliver Beckstein <notifications@github.com>
va escriure:

> Thank you for all the work and extra work addressing the review comments.
> This is a nice addition!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/GridDataFormats/pull/70?email_source=notifications&email_token=AAISB3MC3N46WNXZIAJYYELQSRBW7A5CNFSM4JGYUAA2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDNCHKQ#issuecomment-551166890>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAISB3IVULRWYCUXACKXDTLQSRBW7ANCNFSM4JGYUAAQ>
> .
>
